12-version_control.Rmd

```diff
@@ -0,0 +1,821 @@
+> and me-from-two-months-ago never responds to email.*
+> 
+> --Mark T. Holder
+
+In order to version control a project, you generally need two things: 
+a *version control system* and a *repository hosting service*. 
+The version control system is the software that is responsible 
+for tracking changes, sharing changes you make with others, 
+obtaining changes others have made, and resolving conflicting edits.
+The repository hosting service is responsible for storing a copy 
+of the version controlled project online (a *repository*), 
+where you and your collaborators can access it remotely, 
+discuss issues and bugs, and distribute your final product. 
+For both of these items, there is a wide variety of choices; 
+some of the more popular ones are:
+
+- **Version control systems:**
+    - [Git](https://git-scm.com) 
+    - [Mercurial](https://mercurial-scm.org)
+    - [Subversion](https://subversion.apache.org)
+- **Repository hosting services:**
+    - [GitHub](https://github.com)
+    - [GitLab](https://gitlab.com)
+    - [BitBucket](https://bitbucket.org)
+
+In this textbook we'll use Git for version control, 
+and GitHub for repository hosting, 
+because both are currently the most widely-used platforms.
+
+> **Note:** technically you don't *have to* use a repository hosting service. 
+> You can, for example, use Git to version control a project
+> that is stored only in a folder on your computer. 
+> But using a repository hosting service provides a few big benefits, 
+> including managing collaborator access permissions,
+> tools to discuss and track bugs, 
+> and the ability to have external collaborators contribute work, 
+> not to mention the safety of having your work backed up in the cloud. 
+> Since most repository hosting services now offer free accounts, 
+> there are not many situations in which you wouldn't 
+> want to use one for your project. 
+
+### Version control repositories
+
+Typically when we put a data analysis project under version control, 
+we create two copies of the repository. 
+One we store on our local computer where we most commonly create, 
+edit and delete files.
+This copy is commonly referred to as our **local** repository. The other copy
+is typically stored on the repository hosting service (*e.g.,* GitHub), where
+we can easily share it with our collaborators. 
+This copy is commonly referred to as our **remote** repository 
+(Figure \@ref(fig:vc1-no-changes)).
+
+> **Note:** In this book, we consider a workspace on a server 
+> (*e.g.,* a Jupyterhub), 
+> as well as a laptop or desktop computer to be the local computer.
+
+```{r vc1-no-changes, fig.cap = 'Schematic of local and remote version control repositories.', fig.retina = 2}
+image_read("img/vc2-changes.png") %>%
+  image_crop("3632x2000")
+```
+
+Both copies of the repository have a working directory, 
+where you easily create, edit and delete files, 
+as well as a history (Figure \@ref(fig:vc1-no-changes)). 
```

Comment:
I wonder as well if we change the layout to that the diagrams are going left to right instead since we read left to right? and then make the online stuff above since we often think "cloud" up ... Maybe something like this: 
![git_github 002](https://user-images.githubusercontent.com/3838091/120852569-38158000-c52f-11eb-9a09-c0b81779bb64.jpeg)
Though this is probably a ton of work so I don't think its super necessary